File: data_baseline/openrouter__qwen__qwen3.7-plus/haskell/rep10/Main.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE DeriveGeneric #-}

import Web.Scotty
import qualified Network.HTTP.Types as HT
import Data.Aeson
import GHC.Generics
import qualified Data.Map.Strict as M
import qualified Data.IntMap.Strict as IM
import Control.Concurrent.STM
import Data.Time.Clock
import Data.Time.Format
import Data.UUID.V4
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Lazy as TL
import qualified Data.Text.Encoding as TE
import qualified Data.ByteString.Char8 as BS
import qualified Data.ByteString.Lazy as LBS
import Data.Char (isAlphaNum)
import Crypto.Hash (hash, SHA256, Digest)
import Data.Maybe (fromMaybe)
import Text.Read (readMaybe)
import Data.List (find, sortBy)
import Data.Ord (comparing)
import System.Environment (getArgs)

data User = User
  { userId :: Int
  , username :: Text
  , passwordHash :: Text
  } deriving (Show, Generic)

data Todo = Todo
  { todoId :: Int
  , ownerId :: Int
  , title :: Text
  , description :: Text
  , completed :: Bool
  , createdAt :: UTCTime
  , updatedAt :: UTCTime
  } deriving (Show, Generic)

data AppState = AppState
  { stateUsersByName :: TVar (M.Map Text User)
  , stateUsersById :: TVar (IM.IntMap User)
  , stateNextUserId :: TVar Int
  , stateSessions :: TVar (M.Map Text Int)
  , stateTodos :: TVar (IM.IntMap Todo)
  , stateNextTodoId :: TVar Int
  }

initAppState :: IO AppState
initAppState = do
  usersByName <- newTVarIO M.empty
  usersById <- newTVarIO IM.empty
  nextUserId <- newTVarIO 1
  sessions <- newTVarIO M.empty
  todos <- newTVarIO IM.empty
  nextTodoId <- newTVarIO 1
  return AppState
    { stateUsersByName = usersByName
    , stateUsersById = usersById
    , stateNextUserId = nextUserId
    , stateSessions = sessions
    , stateTodos = todos
    , stateNextTodoId = nextTodoId
    }

hashPassword :: Text -> Text -> Text
hashPassword username pass =
  let salt = "SUPER_SECRET_SALT_123"
      input = TE.encodeUtf8 (salt <> ":" <> username <> ":" <> pass)
      digest = hash input :: Digest SHA256
  in T.pack (show digest)

jsonError :: Int -> Text -> ActionM a
jsonError sc msg = do
  status (HT.mkStatus sc "")
  addHeader "Content-Type" "application/json"
  raw (encode $ object ["error" .= msg])
  finish

jsonWithStatus :: ToJSON a => Int -> a -> ActionM ()
jsonWithStatus sc val = do
  status (HT.mkStatus sc "")
  addHeader "Content-Type" "application/json"
  raw (encode val)

getCookieValue :: Text -> ActionM (Maybe Text)
getCookieValue name = do
  mCookie <- getCookie name
  return mCookie

requireAuth :: AppState -> ActionM Int
requireAuth appState = do
  mToken <- getCookieValue "session_id"
  case mToken of
    Nothing -> jsonError 401 "Authentication required"
    Just token -> do
      sessions <- liftIO $ readTVarIO (stateSessions appState)
      case M.lookup token sessions of
        Nothing -> jsonError 401 "Authentication required"
        Just uid -> return uid

todoToJSON :: Todo -> Value
todoToJSON t = object
  [ "id" .= todoId t
  , "title" .= title t
  , "description" .= description t
  , "completed" .= completed t
  , "created_at" .= formatTimeTodo (createdAt t)
  , "updated_at" .= formatTimeTodo (updatedAt t)
  ]

formatTimeTodo :: UTCTime -> Text
formatTimeTodo t = T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" t

main :: IO ()
main = do
  appState <- initAppState
  portArg <- getPortArg
  scotty portArg $ do
    
    post "/register" $ do
      bodyData <- body
      case eitherDecode bodyData of
        Left _ -> jsonError 400 "Invalid username"
        Right (RegisterReq uname pass) -> do
          if not (T.length uname >= 3 && T.length uname <= 50) then
            jsonError 400 "Invalid username"
          else if not (T.all (\c -> isAlphaNum c || c == '_') uname) then
            jsonError 400 "Invalid username"
          else if T.length pass < 8 then
            jsonError 400 "Password too short"
          else do
            usersMap <- liftIO $ readTVarIO (stateUsersByName appState)
            if M.member uname usersMap then
              jsonError 409 "Username already exists"
            else do
              uid <- liftIO $ atomically $ do
                uid <- readTVar (stateNextUserId appState)
                writeTVar (stateNextUserId appState) (uid + 1)
                let newUser = User uid uname (hashPassword uname pass)
                modifyTVar (stateUsersByName appState) (M.insert uname newUser)
                modifyTVar (stateUsersById appState) (IM.insert uid newUser)
                return uid
              jsonWithStatus 201 $ object ["id" .= uid, "username" .= uname]

    post "/login" $ do
      bodyData <- body
      case eitherDecode bodyData of
        Left _ -> jsonError 401 "Invalid credentials"
        Right (LoginReq uname pass) -> do
          usersMap <- liftIO $ readTVarIO (stateUsersByName appState)
          case M.lookup uname usersMap of
            Nothing -> jsonError 401 "Invalid credentials"
            Just user -> 
              if passwordHash user == hashPassword uname pass then do
                token <- liftIO $ show <$> nextRandom
                liftIO $ atomically $ modifyTVar (stateSessions appState) (M.insert (T.pack token) (userId user))
                addHeader "Set-Cookie" (TL.pack $ "session_id=" ++ token ++ "; Path=/; HttpOnly")
                jsonWithStatus 200 $ object ["id" .= userId user, "username" .= username user]
              else
                jsonError 401 "Invalid credentials"

    post "/logout" $ do
      _ <- requireAuth appState
      mToken <- getCookieValue "session_id"
      case mToken of
        Just token -> liftIO $ atomically $ modifyTVar (stateSessions appState) (M.delete token)
        Nothing -> return ()
      jsonWithStatus 200 (object [])

    get "/me" $ do
      uid <- requireAuth appState
      usersById <- liftIO $ readTVarIO (stateUsersById appState)
      case IM.lookup uid usersById of
        Nothing -> jsonError 500 "User not found"
        Just user -> jsonWithStatus 200 $ object ["id" .= userId user, "username" .= username user]

    put "/password" $ do
      uid <- requireAuth appState
      bodyData <- body
      case eitherDecode bodyData of
        Left _ -> jsonError 400 "Invalid request"
        Right (PasswordReq oldPass newPass) -> do
          if T.length newPass < 8 then
            jsonError 400 "Password too short"
          else do
            usersById <- liftIO $ readTVarIO (stateUsersById appState)
            case IM.lookup uid usersById of
              Nothing -> jsonError 500 "User not found"
              Just user -> do
                if passwordHash user /= hashPassword (username user) oldPass then
                  jsonError 401 "Invalid credentials"
                else do
                  let newHash = hashPassword (username user) newPass
                      newUser = user { passwordHash = newHash }
                  liftIO $ atomically $ do
                    modifyTVar (stateUsersById appState) (IM.insert uid newUser)
                    modifyTVar (stateUsersByName appState) (M.insert (username user) newUser)
                  jsonWithStatus 200 (object [])

    get "/todos" $ do
      uid <- requireAuth appState
      todosMap <- liftIO $ readTVarIO (stateTodos appState)
      let userTodos = filter (\t -> ownerId t == uid) (IM.elems todosMap)
          sortedTodos = sortBy (comparing todoId) userTodos
      jsonWithStatus 200 (map todoToJSON sortedTodos)

    post "/todos" $ do
      uid <- requireAuth appState
      bodyData <- body
      case eitherDecode bodyData of
        Left _ -> jsonError 400 "Title is required"
        Right (CreateTodoReq t mDesc) -> do
          let d = fromMaybe "" mDesc
          if T.null t then
            jsonError 400 "Title is required"
          else do
            now <- liftIO getCurrentTime
            tid <- liftIO $ atomically $ do
              tid <- readTVar (stateNextTodoId appState)
              writeTVar (stateNextTodoId appState) (tid + 1)
              let newTodo = Todo tid uid t d False now now
              modifyTVar (stateTodos appState) (IM.insert tid newTodo)
              return tid
            
            todosMap <- liftIO $ readTVarIO (stateTodos appState)
            case IM.lookup tid todosMap of
              Nothing -> jsonError 500 "Failed to create todo"
              Just newTodo -> jsonWithStatus 201 (todoToJSON newTodo)

    get "/todos/:id" $ do
      uid <- requireAuth appState
      tidStr <- pathParam "id"
      case readMaybe (T.unpack tidStr) :: Maybe Int of
        Nothing -> jsonError 404 "Todo not found"
        Just tid -> do
          todosMap <- liftIO $ readTVarIO (stateTodos appState)
          case IM.lookup tid todosMap of
            Nothing -> jsonError 404 "Todo not found"
            Just todo -> 
              if ownerId todo /= uid then
                jsonError 404 "Todo not found"
              else
                jsonWithStatus 200 (todoToJSON todo)

    put "/todos/:id" $ do
      uid <- requireAuth appState
      tidStr <- pathParam "id"
      case readMaybe (T.unpack tidStr) :: Maybe Int of
        Nothing -> jsonError 404 "Todo not found"
        Just tid -> do
          bodyData <- body
          let parsed = if LBS.null bodyData 
                       then Right (UpdateTodoReq Nothing Nothing Nothing) 
                       else eitherDecode bodyData :: Either String UpdateTodoReq
          case parsed of
            Left _ -> jsonError 400 "Invalid request"
            Right req -> do
              todosMap <- liftIO $ readTVarIO (stateTodos appState)
              case IM.lookup tid todosMap of
                Nothing -> jsonError 404 "Todo not found"
                Just todo -> 
                  if ownerId todo /= uid then
                    jsonError 404 "Todo not found"
                  else do
                    let newTitle = fromMaybe (title todo) (updTitle req)
                    if T.null newTitle then
                      jsonError 400 "Title is required"
                    else do
                      now <- liftIO getCurrentTime
                      let newDesc = fromMaybe (description todo) (updDescription req)
                          newCompleted = fromMaybe (completed todo) (updCompleted req)
                          updatedTodo = todo 
                            { title = newTitle
                            , description = newDesc
                            , completed = newCompleted
                            , updatedAt = now
                            }
                      liftIO $ atomically $ modifyTVar (stateTodos appState) (IM.insert tid updatedTodo)
                      jsonWithStatus 200 (todoToJSON updatedTodo)

    delete "/todos/:id" $ do
      uid <- requireAuth appState
      tidStr <- pathParam "id"
      case readMaybe (T.unpack tidStr) :: Maybe Int of
        Nothing -> jsonError 404 "Todo not found"
        Just tid -> do
          todosMap <- liftIO $ readTVarIO (stateTodos appState)
          case IM.lookup tid todosMap of
            Nothing -> jsonError 404 "Todo not found"
            Just todo -> 
              if ownerId todo /= uid then
                jsonError 404 "Todo not found"
              else do
                liftIO $ atomically $ modifyTVar (stateTodos appState) (IM.delete tid)
                status (HT.mkStatus 204 "")
                raw ""

data RegisterReq = RegisterReq { regUsername :: Text, regPassword :: Text }
instance FromJSON RegisterReq where
  parseJSON = withObject "RegisterReq" $ \v -> RegisterReq
    <$> v .: "username"
    <*> v .: "password"

data LoginReq = LoginReq { logUsername :: Text, logPassword :: Text }
instance FromJSON LoginReq where
  parseJSON = withObject "LoginReq" $ \v -> LoginReq
    <$> v .: "username"
    <*> v .: "password"

data PasswordReq = PasswordReq { pwdOldPassword :: Text, pwdNewPassword :: Text }
instance FromJSON PasswordReq where
  parseJSON = withObject "PasswordReq" $ \v -> PasswordReq
    <$> v .: "old_password"
    <*> v .: "new_password"

data CreateTodoReq = CreateTodoReq { todoTitle :: Text, todoDescription :: Maybe Text }
instance FromJSON CreateTodoReq where
  parseJSON = withObject "CreateTodoReq" $ \v -> CreateTodoReq
    <$> v .: "title"
    <*> v .:? "description"

data UpdateTodoReq = UpdateTodoReq 
  { updTitle :: Maybe Text
  , updDescription :: Maybe Text
  , updCompleted :: Maybe Bool
  }
instance FromJSON UpdateTodoReq where
  parseJSON = withObject "UpdateTodoReq" $ \v -> UpdateTodoReq
    <$> v .:? "title"
    <*> v .:? "description"
    <*> v .:? "completed"

getPortArg :: IO Int
getPortArg = do
  args <- getArgs
  let port = case args of
        ["--port", p] -> read p
        _ -> 8080
  return port

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep10/src/main.rs
use axum::{
    extract::{Path, State},
    http::{header, HeaderMap, StatusCode},
    response::{IntoResponse, Response},
    routing::{get, post, put, delete},
    Json, Router,
};
use chrono::Utc;
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU32, Ordering};
use std::sync::{Arc, OnceLock};
use tokio::sync::RwLock;
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct User {
    pub id: u32,
    pub username: String,
    pub password: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Todo {
    pub id: u32,
    pub user_id: u32,
    pub title: String,
    pub description: String,
    pub completed: bool,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Clone)]
pub struct AppState {
    pub users: Arc<RwLock<HashMap<u32, User>>>,
    pub sessions: Arc<RwLock<HashMap<String, u32>>>,
    pub todos: Arc<RwLock<HashMap<u32, Todo>>>,
    pub next_user_id: Arc<AtomicU32>,
    pub next_todo_id: Arc<AtomicU32>,
}

fn username_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"^[a-zA-Z0-9_]{3,50}$").unwrap())
}

async fn require_auth(
    state: &AppState,
    headers: &HeaderMap,
) -> Result<u32, Response> {
    let cookie_header = headers.get(header::COOKIE).and_then(|v| v.to_str().ok());
    let session_id = cookie_header
        .and_then(|c| {
            c.split(';')
                .find(|p| p.trim().starts_with("session_id="))
                .map(|p| p.trim().trim_start_matches("session_id=").to_string())
        });
    
    if let Some(sid) = session_id {
        let sessions = state.sessions.read().await;
        if let Some(&user_id) = sessions.get(&sid) {
            return Ok(user_id);
        }
    }
    Err((
        StatusCode::UNAUTHORIZED,
        Json(json!({"error": "Authentication required"})),
    ).into_response())
}

#[derive(Deserialize)]
struct RegisterRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn register(
    State(state): State<Arc<AppState>>,
    Json(req): Json<RegisterRequest>,
) -> impl IntoResponse {
    let username = match req.username {
        Some(u) if username_re().is_match(&u) => u,
        _ => return (StatusCode::BAD_REQUEST, Json(json!({"error": "Invalid username"}))).into_response(),
    };
    
    let password = match req.password {
        Some(p) if p.len() >= 8 => p,
        _ => return (StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))).into_response(),
    };

    let mut users = state.users.write().await;
    if users.values().any(|u| u.username == username) {
        return (StatusCode::CONFLICT, Json(json!({"error": "Username already exists"}))).into_response();
    }

    let user_id = state.next_user_id.fetch_add(1, Ordering::SeqCst);
    users.insert(user_id, User {
        id: user_id,
        username: username.clone(),
        password,
    });

    (StatusCode::CREATED, Json(json!({"id": user_id, "username": username}))).into_response()
}

#[derive(Deserialize)]
struct LoginRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn login(
    State(state): State<Arc<AppState>>,
    Json(req): Json<LoginRequest>,
) -> impl IntoResponse {
    let username = req.username.unwrap_or_default();
    let password = req.password.unwrap_or_default();

    let user_id = {
        let users = state.users.read().await;
        users.values().find(|u| u.username == username && u.password == password).map(|u| u.id)
    };

    let user_id = match user_id {
        Some(id) => id,
        None => return (StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))).into_response(),
    };

    let session_id = Uuid::new_v4().to_string();
    {
        let mut sessions = state.sessions.write().await;
        sessions.insert(session_id.clone(), user_id);
    }

    let users = state.users.read().await;
    let user = users.get(&user_id).unwrap();

    let mut headers = HeaderMap::new();
    let cookie_val = format!("session_id={}; Path=/; HttpOnly", session_id);
    headers.insert(header::SET_COOKIE, cookie_val.parse().unwrap());

    (StatusCode::OK, headers, Json(json!({"id": user.id, "username": user.username}))).into_response()
}

async fn logout(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> impl IntoResponse {
    let _ = require_auth(&state, &headers).await; // Just to verify auth

    let cookie_header = headers.get(header::COOKIE).and_then(|v| v.to_str().ok());
    if let Some(sid) = cookie_header.and_then(|c| {
        c.split(';')
            .find(|p| p.trim().starts_with("session_id="))
            .map(|p| p.trim().trim_start_matches("session_id=").to_string())
    }) {
        let mut sessions = state.sessions.write().await;
        sessions.remove(&sid);
    }

    (StatusCode::OK, Json(json!({}))).into_response()
}

async fn me(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let users = state.users.read().await;
    if let Some(user) = users.get(&user_id) {
        (StatusCode::OK, Json(json!({"id": user.id, "username": user.username}))).into_response()
    } else {
        (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))).into_response()
    }
}

#[derive(Deserialize)]
struct PasswordRequest {
    old_password: Option<String>,
    new_password: Option<String>,
}

async fn change_password(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(req): Json<PasswordRequest>,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let old_password = req.old_password.unwrap_or_default();
    let new_password = req.new_password.unwrap_or_default();

    if new_password.len() < 8 {
        return (StatusCode::BAD_REQUEST, Json(json!({"error": "Password too short"}))).into_response();
    }

    let mut users = state.users.write().await;
    if let Some(user) = users.get_mut(&user_id) {
        if user.password != old_password {
            return (StatusCode::UNAUTHORIZED, Json(json!({"error": "Invalid credentials"}))).into_response();
        }
        user.password = new_password;
        (StatusCode::OK, Json(json!({}))).into_response()
    } else {
        (StatusCode::UNAUTHORIZED, Json(json!({"error": "Authentication required"}))).into_response()
    }
}

async fn list_todos(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let todos = state.todos.read().await;
    let mut user_todos: Vec<_> = todos.values()
        .filter(|t| t.user_id == user_id)
        .collect();
    
    user_todos.sort_by_key(|t| t.id);
    
    (StatusCode::OK, Json(user_todos)).into_response()
}

#[derive(Deserialize)]
struct CreateTodoRequest {
    title: Option<String>,
    description: Option<String>,
}

async fn create_todo(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Json(req): Json<CreateTodoRequest>,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let title = match req.title {
        Some(t) if !t.is_empty() => t,
        _ => return (StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))).into_response(),
    };

    let description = req.description.unwrap_or_default();
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
    let todo_id = state.next_todo_id.fetch_add(1, Ordering::SeqCst);

    let todo = Todo {
        id: todo_id,
        user_id,
        title,
        description,
        completed: false,
        created_at: now.clone(),
        updated_at: now,
    };

    state.todos.write().await.insert(todo_id, todo.clone());
    (StatusCode::CREATED, Json(todo)).into_response()
}

async fn get_todo(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Path(todo_id): Path<u32>,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let todos = state.todos.read().await;
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id == user_id {
            return (StatusCode::OK, Json(todo.clone())).into_response();
        }
    }
    (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))).into_response()
}

#[derive(Deserialize)]
struct UpdateTodoRequest {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

async fn update_todo(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Path(todo_id): Path<u32>,
    Json(req): Json<UpdateTodoRequest>,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let mut todos = state.todos.write().await;
    if let Some(todo) = todos.get_mut(&todo_id) {
        if todo.user_id != user_id {
            return (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))).into_response();
        }

        if let Some(title) = req.title {
            if title.is_empty() {
                return (StatusCode::BAD_REQUEST, Json(json!({"error": "Title is required"}))).into_response();
            }
            todo.title = title;
        }

        if let Some(desc) = req.description {
            todo.description = desc;
        }

        if let Some(completed) = req.completed {
            todo.completed = completed;
        }

        todo.updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();
        
        return (StatusCode::OK, Json(todo.clone())).into_response();
    }
    (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))).into_response()
}

async fn delete_todo(
    State(state): State<Arc<AppState>>,
    headers: HeaderMap,
    Path(todo_id): Path<u32>,
) -> impl IntoResponse {
    let user_id = match require_auth(&state, &headers).await {
        Ok(id) => id,
        Err(resp) => return resp,
    };

    let mut todos = state.todos.write().await;
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id != user_id {
            return (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))).into_response();
        }
        todos.remove(&todo_id);
        return StatusCode::NO_CONTENT.into_response();
    }
    (StatusCode::NOT_FOUND, Json(json!({"error": "Todo not found"}))).into_response()
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut port = 3000;
    
    let mut i = 1;
    while i < args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse().expect("Invalid port number");
            i += 2;
        } else {
            i += 1;
        }
    }

    let state = Arc::new(AppState {
        users: Arc::new(RwLock::new(HashMap::new())),
        sessions: Arc::new(RwLock::new(HashMap::new())),
        todos: Arc::new(RwLock::new(HashMap::new())),
        next_user_id: Arc::new(AtomicU32::new(1)),
        next_todo_id: Arc::new(AtomicU32::new(1)),
    });

    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(me))
        .route("/password", put(change_password))
        .route("/todos", get(list_todos))
        .route("/todos", post(create_todo))
        .route("/todos/{id}", get(get_todo))
        .route("/todos/{id}", put(update_todo))
        .route("/todos/{id}", delete(delete_todo))
        .with_state(state);

    let addr = format!("0.0.0.0:{}", port);
    println!("Starting server on {}", addr);
    let listener = tokio::net::TcpListener::bind(&addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
